planet/auth.py

```diff
@@ -0,0 +1,285 @@
+            password: Planet account password.
+            base_url: The base URL to use. Defaults to production
+                authentication API base url.
+        '''
+        cl = AuthClient(base_url=base_url)
+        auth_data = cl.login(email, password)
+
+        api_key = auth_data['api_key']
+        auth = APIKeyAuth(api_key)
+        LOGGER.debug('Auth set from login email and password')
+        return auth
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(
+        cls,
+        data: dict
+    ) -> Auth:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def value(self):
+        pass
+
+    def write(
+        self,
+        filename: str = None
+    ):
+        '''Write authentication information.
+
+        Parameters:
+            filename: Alternate path for the planet secret file.
+        '''
+        filename = filename or SECRET_FILE_PATH
+        secret_file = _SecretFile(filename)
+        secret_file.write(self.to_dict())
+
+
+class AuthClient():
+    def __init__(
+        self,
+        base_url: str = None
+    ):
+        """
+        Parameters:
+            base_url: Alternate authentication api base URL.
+        """
+        self._base_url = base_url or AUTH_URL
+        if not self._base_url.endswith('/'):
+            self._base_url += '/'
+
+    def login(
+        self,
+        email: str,
+        password: str
+    ) -> dict:
+        '''Login using email identity and credentials.
+
+        Note: To keep your password secure, the use of `getpass` is
+        recommended.
+
+        Parameters:
+            email: Planet account email address.
+            password:  Planet account password.
+
+        Returns:
+             A JSON object containing an `api_key` property with the user's
+        API_KEY.
+        '''
+        url = self._base_url + 'login'
+        data = {'email': email,
+                'password': password
+                }
+
+        sess = http.AuthSession()
+        req = models.Request(url, method='POST', json=data)
+        resp = sess.request(req)
+        auth_data = self.decode_response(resp)
+        return auth_data
+
+    @staticmethod
+    def decode_response(response):
+        '''This is magic I don't understand'''
+        jwt = response.json()['token']
+
+        # stuff before the first '.' and after the second '.' doesn't matter
+        payload = jwt.split('.')[1]
+
+        # the '===' addition ensures adequate padding
+        payload = base64.urlsafe_b64decode(payload + '===')
+        decoded = json.loads(payload.decode())
+        return decoded
+
+
+class APIKeyAuthException(Exception):
+    '''exceptions thrown by APIKeyAuth'''
+    pass
+
+
+class APIKeyAuth(httpx.BasicAuth, Auth):
+    '''Planet API Key authentication.'''
+    DICT_KEY = 'key'
+
+    def __init__(
+        self,
+        key: str
+    ):
+        '''Initialize APIKeyAuth.
+
+        Parameters:
+            key: API key.
```

Comment:
why not call this `api_key` rather than `key` to be more explicit?

Reply:
Mostly for succinctness. Since the class is called `APIKeyAuth`, it felt redundant to use `api_key`.